File: harness/hardware.py
"""Concrete virtual hardware profiles for CodexOS generations."""

from __future__ import annotations

import json
import os
import re
import subprocess
from dataclasses import dataclass
from pathlib import Path

_PROFILE_PATTERN = re.compile(r"[a-z0-9][a-z0-9-]{0,63}")
_CPU_PATTERN = re.compile(r"[A-Za-z0-9_.+-]{1,64}")
_QEMU_VERSION_LIMIT = 256


@dataclass(frozen=True)
class HardwareManifest:
    schema_version: int
    profile: str
    machine: str
    accelerator: str
    cpu_model: str
    vcpus: int
    memory_mib: int
    graphics: str
    network: str
    writable_block_devices: tuple[str, ...]
    qemu_version: str
    qemu_arguments: tuple[str, ...]

    def as_json_object(self) -> dict[str, object]:
        return {
            "schema_version": self.schema_version,
            "profile": self.profile,
            "machine": self.machine,
            "accelerator": self.accelerator,
            "cpu_model": self.cpu_model,
            "vcpus": self.vcpus,
            "memory_mib": self.memory_mib,
            "graphics": self.graphics,
            "network": self.network,
            "writable_block_devices": list(self.writable_block_devices),
            "qemu_version": self.qemu_version,
            "qemu_arguments": list(self.qemu_arguments),
        }


@dataclass(frozen=True)
class CodexOSHardwareProfile:
    """One trusted, fixed QEMU hardware configuration."""

    profile: str
    machine: str
    accelerator: str
    cpu_model: str
    vcpus: int
    memory_mib: int
    graphics: str = "std-vga"
    network: str = "none"
    writable_block_devices: tuple[str, ...] = ()

    def __post_init__(self) -> None:
        _validate_profile_values(
            self.profile,
            self.machine,
            self.accelerator,
            self.cpu_model,
            self.vcpus,
            self.memory_mib,
            self.graphics,
            self.network,
            self.writable_block_devices,
        )

    def require_available(self) -> None:
        if self.accelerator != "kvm":
            return
        kvm = Path("/dev/kvm")
        if not kvm.exists() or not os.access(kvm, os.R_OK | os.W_OK):
            raise RuntimeError(
                f"{self.profile} requires KVM, but /dev/kvm is unavailable"
            )

    def qemu_arguments(
        self,
        boot_iso: Path,
        qmp_socket: Path,
        serial_socket: Path,
    ) -> list[str]:
        boot_file = json.dumps(
            {
                "driver": "file",
                "filename": str(boot_iso),
                "node-name": "codexos-boot-file",
                "read-only": True,
            },
            separators=(",", ":"),
        )
        boot_raw = json.dumps(
            {
                "driver": "raw",
                "file": "codexos-boot-file",
                "node-name": "codexos-boot-cd",
                "read-only": True,
            },
            separators=(",", ":"),
        )
        return [
            "-machine",
            (
                f"{self.machine},accel={self.accelerator},"
                "pcspk-audiodev=codexos-noaudio"
            ),
            "-cpu",
            self.cpu_model,
            "-smp",
            str(self.vcpus),
            "-m",
            f"{self.memory_mib}M",
            "-nodefaults",
            "-display",
            "none",
            "-monitor",
            "none",
            "-no-reboot",
            "-nic",
            "none",
            "-audiodev",
            "none,id=codexos-noaudio",
            "-blockdev",
            boot_file,
            "-blockdev",
            boot_raw,
            "-device",
            "ide-cd,drive=codexos-boot-cd,bootindex=1",
            "-device",
            "VGA",
            "-chardev",
            (
                "socket,id=codexos-com1,"
                f"path={serial_socket},server=on,wait=off"
            ),
            "-device",
            "isa-serial,chardev=codexos-com1,index=0",
            "-qmp",
            f"unix:{qmp_socket},server=on,wait=off",
        ]

    def manifest(self, qemu_version: str) -> HardwareManifest:
        normalized_arguments = self.qemu_arguments(
            Path("<BOOT_ISO>"),
            Path("<QMP_SOCKET>"),
            Path("<SERIAL_SOCKET>"),
        )
        return validate_hardware_manifest(
            {
                "schema_version": 1,
                "profile": self.profile,
                "machine": self.machine,
                "accelerator": self.accelerator,
                "cpu_model": self.cpu_model,
                "vcpus": self.vcpus,
                "memory_mib": self.memory_mib,
                "graphics": self.graphics,
                "network": self.network,
                "writable_block_devices": list(
                    self.writable_block_devices
                ),
                "qemu_version": qemu_version,
                "qemu_arguments": normalized_arguments,
            }
        )


def discover_qemu_version(executable: str) -> str:
    try:
        result = subprocess.run(
            [executable, "--version"],
            stdout=subprocess.PIPE,
            stderr=subprocess.STDOUT,
            text=True,
            timeout=5.0,
            check=False,
        )
    except (OSError, subprocess.TimeoutExpired) as error:
        raise RuntimeError(f"could not determine QEMU version: {error}") from error
    if result.returncode != 0:
        raise RuntimeError("could not determine QEMU version")
    lines = result.stdout.splitlines()
    if not lines:
        raise RuntimeError("QEMU version output is empty")
    version = lines[0]
    if (
        not version
        or len(version.encode("utf-8")) > _QEMU_VERSION_LIMIT
        or not version.isprintable()
    ):
        raise RuntimeError("QEMU version output is invalid")
    return version


def validate_hardware_manifest(value: object) -> HardwareManifest:
    expected = {
        "schema_version",
        "profile",
        "machine",
        "accelerator",
        "cpu_model",
        "vcpus",
        "memory_mib",
        "graphics",
        "network",
        "writable_block_devices",
        "qemu_version",
        "qemu_arguments",
    }
    if not isinstance(value, dict) or set(value) != expected:
        raise ValueError("generation hardware manifest is malformed")
    if type(value["schema_version"]) is not int or value["schema_version"] != 1:
        raise ValueError("generation hardware manifest is malformed")

    writable = value["writable_block_devices"]
    if not isinstance(writable, list) or writable:
        raise ValueError("generation hardware manifest is malformed")
    _validate_profile_values(
        value["profile"],
        value["machine"],
        value["accelerator"],
        value["cpu_model"],
        value["vcpus"],
        value["memory_mib"],
        value["graphics"],
        value["network"],
        tuple(writable),
    )
    qemu_version = value["qemu_version"]
    if (
        not isinstance(qemu_version, str)
        or not qemu_version
        or len(qemu_version.encode("utf-8")) > _QEMU_VERSION_LIMIT
        or not qemu_version.isprintable()
    ):
        raise ValueError("generation hardware manifest is malformed")
    qemu_arguments = value["qemu_arguments"]
    profile = CodexOSHardwareProfile(
        profile=value["profile"],
        machine=value["machine"],
        accelerator=value["accelerator"],
        cpu_model=value["cpu_model"],
        vcpus=value["vcpus"],
        memory_mib=value["memory_mib"],
        graphics=value["graphics"],
        network=value["network"],
        writable_block_devices=tuple(writable),
    )
    expected_arguments = profile.qemu_arguments(
        Path("<BOOT_ISO>"),
        Path("<QMP_SOCKET>"),
        Path("<SERIAL_SOCKET>"),
    )
    if qemu_arguments != expected_arguments:
        raise ValueError("generation hardware manifest is malformed")
    return HardwareManifest(
        schema_version=1,
        profile=value["profile"],
        machine=value["machine"],
        accelerator=value["accelerator"],
        cpu_model=value["cpu_model"],
        vcpus=value["vcpus"],
        memory_mib=value["memory_mib"],
        graphics=value["graphics"],
        network=value["network"],
        writable_block_devices=tuple(writable),
        qemu_version=qemu_version,
        qemu_arguments=tuple(qemu_arguments),
    )


def _validate_profile_values(
    profile: object,
    machine: object,
    accelerator: object,
    cpu_model: object,
    vcpus: object,
    memory_mib: object,
    graphics: object,
    network: object,
    writable_block_devices: object,
) -> None:
    if not isinstance(profile, str) or _PROFILE_PATTERN.fullmatch(profile) is None:
        raise ValueError("invalid CodexOS hardware profile")
    if machine != "q35":
        raise ValueError("invalid CodexOS hardware machine")
    if accelerator not in {"kvm", "kvm:tcg"}:
        raise ValueError("invalid CodexOS hardware accelerator")
    if not isinstance(cpu_model, str) or _CPU_PATTERN.fullmatch(cpu_model) is None:
        raise ValueError("invalid CodexOS hardware CPU model")
    if type(vcpus) is not int or not 1 <= vcpus <= 256:
        raise ValueError("invalid CodexOS hardware vCPU count")
    if type(memory_mib) is not int or not 1 <= memory_mib <= 1_048_576:
        raise ValueError("invalid CodexOS hardware memory size")
    if graphics != "std-vga" or network != "none":
        raise ValueError("invalid CodexOS peripheral hardware")
    if writable_block_devices != ():
        raise ValueError("writable block devices are not supported")


EXPERIMENT_HARDWARE_PROFILE = CodexOSHardwareProfile(
    profile="experiment-v1",
    machine="q35",
    accelerator="kvm",
    cpu_model="host",
    vcpus=4,
    memory_mib=8192,
)

TEST_HARDWARE_PROFILE = CodexOSHardwareProfile(
    profile="test-v1",
    machine="q35",
    accelerator="kvm:tcg",
    cpu_model="qemu64",
    vcpus=1,
    memory_mib=128,
)

File: internal/qemu/hardware.go
package qemu

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"os/exec"
	"regexp"
	"strings"
	"syscall"
	"time"
	"unicode"
	"unicode/utf16"
	"unicode/utf8"
)

const (
	hardwareSchemaVersion = 1
	qemuVersionLimit      = 256
	qemuVersionOutputMax  = 1024 * 1024
	hardwareManifestMax   = 1024 * 1024
	kvmReadWriteAccess    = 6 // POSIX R_OK | W_OK.
)

var (
	hardwareProfilePattern = regexp.MustCompile(`^[a-z0-9][a-z0-9-]{0,63}$`)
	hardwareCPUPattern     = regexp.MustCompile(`^[A-Za-z0-9_.+\-]{1,64}$`)
)

type HardwareError struct {
	Reason string
	Err    error
}

func (e *HardwareError) Error() string {
	if e.Err == nil {
		return e.Reason
	}
	return e.Reason + ": " + e.Err.Error()
}

func (e *HardwareError) Unwrap() error { return e.Err }

type HardwareProfile struct {
	Profile              string
	Machine              string
	Accelerator          string
	CPUModel             string
	VCPUs                int
	MemoryMiB            int
	Graphics             string
	Network              string
	WritableBlockDevices []string
}

type HardwareManifest struct {
	SchemaVersion        int      `json:"schema_version"`
	Profile              string   `json:"profile"`
	Machine              string   `json:"machine"`
	Accelerator          string   `json:"accelerator"`
	CPUModel             string   `json:"cpu_model"`
	VCPUs                int      `json:"vcpus"`
	MemoryMiB            int      `json:"memory_mib"`
	Graphics             string   `json:"graphics"`
	Network              string   `json:"network"`
	WritableBlockDevices []string `json:"writable_block_devices"`
	QEMUVersion          string   `json:"qemu_version"`
	QEMUArguments        []string `json:"qemu_arguments"`
}

var ExperimentHardwareProfile = HardwareProfile{
	Profile:              "experiment-v1",
	Machine:              "q35",
	Accelerator:          "kvm",
	CPUModel:             "host",
	VCPUs:                4,
	MemoryMiB:            8192,
	Graphics:             "std-vga",
	Network:              "none",
	WritableBlockDevices: []string{},
}

var TestHardwareProfile = HardwareProfile{
	Profile:              "test-v1",
	Machine:              "q35",
	Accelerator:          "kvm:tcg",
	CPUModel:             "qemu64",
	VCPUs:                1,
	MemoryMiB:            128,
	Graphics:             "std-vga",
	Network:              "none",
	WritableBlockDevices: []string{},
}

func (p HardwareProfile) Validate() error {
	switch {
	case !hardwareProfilePattern.MatchString(p.Profile):
		return &HardwareError{Reason: "invalid CodexOS hardware profile"}
	case p.Machine != "q35":
		return &HardwareError{Reason: "invalid CodexOS hardware machine"}
	case p.Accelerator != "kvm" && p.Accelerator != "kvm:tcg":
		return &HardwareError{Reason: "invalid CodexOS hardware accelerator"}
	case !hardwareCPUPattern.MatchString(p.CPUModel):
		return &HardwareError{Reason: "invalid CodexOS hardware CPU model"}
	case p.VCPUs < 1 || p.VCPUs > 256:
		return &HardwareError{Reason: "invalid CodexOS hardware vCPU count"}
	case p.MemoryMiB < 1 || p.MemoryMiB > 1_048_576:
		return &HardwareError{Reason: "invalid CodexOS hardware memory size"}
	case p.Graphics != "std-vga" || p.Network != "none":
		return &HardwareError{Reason: "invalid CodexOS peripheral hardware"}
	case len(p.WritableBlockDevices) != 0:
		return &HardwareError{Reason: "writable block devices are not supported"}
	default:
		return nil
	}
}

func (p HardwareProfile) RequireAvailable() error {
	if err := p.Validate(); err != nil {
		return err
	}
	if p.Accelerator != "kvm" {
		return nil
	}
	if err := syscall.Access("/dev/kvm", kvmReadWriteAccess); err != nil {
		return &HardwareError{Reason: fmt.Sprintf("%s requires KVM, but /dev/kvm is unavailable", p.Profile)}
	}
	return nil
}

func (p HardwareProfile) QEMUCommandArguments(bootISO, qmpSocket, serialSocket string) ([]string, error) {
	if err := p.Validate(); err != nil {
		return nil, err
	}
	if !utf8.ValidString(bootISO) || !utf8.ValidString(qmpSocket) || !utf8.ValidString(serialSocket) {
		return nil, &HardwareError{Reason: "QEMU path is not valid UTF-8"}
	}
	bootFile, err := compactASCIIJSON(struct {
		Driver   string `json:"driver"`
		Filename string `json:"filename"`
		NodeName string `json:"node-name"`
		ReadOnly bool   `json:"read-only"`
	}{Driver: "file", Filename: bootISO, NodeName: "codexos-boot-file", ReadOnly: true})
	if err != nil {
		return nil, &HardwareError{Reason: "could not encode QEMU boot file", Err: err}
	}
	bootRaw, err := compactASCIIJSON(struct {
		Driver   string `json:"driver"`
		File     string `json:"file"`
		NodeName string `json:"node-name"`
		ReadOnly bool   `json:"read-only"`
	}{Driver: "raw", File: "codexos-boot-file", NodeName: "codexos-boot-cd", ReadOnly: true})
	if err != nil {
		return nil, &HardwareError{Reason: "could not encode QEMU boot device", Err: err}
	}
	return []string{
		"-machine", p.Machine + ",accel=" + p.Accelerator + ",pcspk-audiodev=codexos-noaudio",
		"-cpu", p.CPUModel,
		"-smp", fmt.Sprintf("%d", p.VCPUs),
		"-m", fmt.Sprintf("%dM", p.MemoryMiB),
		"-nodefaults",
		"-display", "none",
		"-monitor", "none",
		"-no-reboot",
		"-nic", "none",
		"-audiodev", "none,id=codexos-noaudio",
		"-blockdev", bootFile,
		"-blockdev", bootRaw,
		"-device", "ide-cd,drive=codexos-boot-cd,bootindex=1",
		"-device", "VGA",
		"-chardev", "socket,id=codexos-com1,path=" + serialSocket + ",server=on,wait=off",
		"-device", "isa-serial,chardev=codexos-com1,index=0",
		"-qmp", "unix:" + qmpSocket + ",server=on,wait=off",
	}, nil
}

func (p HardwareProfile) Manifest(qemuVersion string) (HardwareManifest, error) {
	arguments, err := p.QEMUCommandArguments("<BOOT_ISO>", "<QMP_SOCKET>", "<SERIAL_SOCKET>")
	if err != nil {
		return HardwareManifest{}, err
	}
	manifest := HardwareManifest{
		SchemaVersion:        hardwareSchemaVersion,
		Profile:              p.Profile,
		Machine:              p.Machine,
		Accelerator:          p.Accelerator,
		CPUModel:             p.CPUModel,
		VCPUs:                p.VCPUs,
		MemoryMiB:            p.MemoryMiB,
		Graphics:             p.Graphics,
		Network:              p.Network,
		WritableBlockDevices: []string{},
		QEMUVersion:          qemuVersion,
		QEMUArguments:        arguments,
	}
	if err := ValidateHardwareManifest(manifest); err != nil {
		return HardwareManifest{}, err
	}
	return manifest, nil
}

func ValidateHardwareManifest(manifest HardwareManifest) error {
	if manifest.SchemaVersion != hardwareSchemaVersion || manifest.WritableBlockDevices == nil || manifest.QEMUArguments == nil {
		return malformedHardwareManifest()
	}
	profile := HardwareProfile{
		Profile:              manifest.Profile,
		Machine:              manifest.Machine,
		Accelerator:          manifest.Accelerator,
		CPUModel:             manifest.CPUModel,
		VCPUs:                manifest.VCPUs,
		MemoryMiB:            manifest.MemoryMiB,
		Graphics:             manifest.Graphics,
		Network:              manifest.Network,
		WritableBlockDevices: manifest.WritableBlockDevices,
	}
	if err := profile.Validate(); err != nil {
		return err
	}
	if !validQEMUVersion(manifest.QEMUVersion) {
		return malformedHardwareManifest()
	}
	expected, err := profile.QEMUCommandArguments("<BOOT_ISO>", "<QMP_SOCKET>", "<SERIAL_SOCKET>")
	if err != nil {
		return err
	}
	if !equalStrings(manifest.QEMUArguments, expected) {
		return malformedHardwareManifest()
	}
	return nil
}

func ParseHardwareManifest(encoded []byte) (HardwareManifest, error) {
	if len(encoded) > hardwareManifestMax || !utf8.Valid(encoded) {
		return HardwareManifest{}, malformedHardwareManifest()
	}
	decoder := json.NewDecoder(bytes.NewReader(encoded))
	var fields map[string]json.RawMessage
	if err := decoder.Decode(&fields); err != nil {
		return HardwareManifest{}, malformedHardwareManifestWith(err)
	}
	if err := requireHardwareJSONEOF(decoder); err != nil {
		return HardwareManifest{}, malformedHardwareManifestWith(err)
	}
	expectedKeys := []string{
		"schema_version", "profile", "machine", "accelerator", "cpu_model", "vcpus",
		"memory_mib", "graphics", "network", "writable_block_devices", "qemu_version", "qemu_arguments",
	}
	if len(fields) != len(expectedKeys) {
		return HardwareManifest{}, malformedHardwareManifest()
	}
	for _, key := range expectedKeys {
		if _, exists := fields[key]; !exists {
			return HardwareManifest{}, malformedHardwareManifest()
		}
	}
	var manifest HardwareManifest
	if err := json.Unmarshal(encoded, &manifest); err != nil {
		return HardwareManifest{}, malformedHardwareManifestWith(err)
	}
	if err := ValidateHardwareManifest(manifest); err != nil {
		return HardwareManifest{}, err
	}
	return manifest, nil
}

func EncodeHardwareManifest(manifest HardwareManifest) ([]byte, error) {
	if err := ValidateHardwareManifest(manifest); err != nil {
		return nil, err
	}
	value := struct {
		Accelerator          string   `json:"accelerator"`
		CPUModel             string   `json:"cpu_model"`
		Graphics             string   `json:"graphics"`
		Machine              string   `json:"machine"`
		MemoryMiB            int      `json:"memory_mib"`
		Network              string   `json:"network"`
		Profile              string   `json:"profile"`
		QEMUArguments        []string `json:"qemu_arguments"`
		QEMUVersion          string   `json:"qemu_version"`
		SchemaVersion        int      `json:"schema_version"`
		VCPUs                int      `json:"vcpus"`
		WritableBlockDevices []string `json:"writable_block_devices"`
	}{
		Accelerator: manifest.Accelerator, CPUModel: manifest.CPUModel, Graphics: manifest.Graphics,
		Machine: manifest.Machine, MemoryMiB: manifest.MemoryMiB, Network: manifest.Network,
		Profile: manifest.Profile, QEMUArguments: manifest.QEMUArguments, QEMUVersion: manifest.QEMUVersion,
		SchemaVersion: manifest.SchemaVersion, VCPUs: manifest.VCPUs, WritableBlockDevices: manifest.WritableBlockDevices,
	}
	var output bytes.Buffer
	encoder := json.NewEncoder(&output)
	encoder.SetEscapeHTML(false)
	encoder.SetIndent("", "  ")
	if err := encoder.Encode(value); err != nil {
		return nil, &HardwareError{Reason: "could not encode generation hardware manifest", Err: err}
	}
	return hardwareASCIIJSON(output.Bytes()), nil
}

func DiscoverQEMUVersion(ctx context.Context, executable string) (string, error) {
	operationContext, cancel := context.WithTimeout(ctx, 5*time.Second)
	defer cancel()
	command := exec.CommandContext(operationContext, executable, "--version")
	output := &boundedOutput{limit: qemuVersionOutputMax}
	command.Stdout = output
	command.Stderr = output
	if err := command.Run(); err != nil {
		if contextErr := operationContext.Err(); contextErr != nil {
			return "", &HardwareError{Reason: "could not determine QEMU version", Err: contextErr}
		}
		return "", &HardwareError{Reason: "could not determine QEMU version", Err: err}
	}
	if output.overflow {
		return "", &HardwareError{Reason: "could not determine QEMU version: output exceeds 1 MiB"}
	}
	if !utf8.Valid(output.bytes) {
		return "", &HardwareError{Reason: "QEMU version output is invalid"}
	}
	version, found := firstPythonTextLine(string(output.bytes))
	if !found {
		return "", &HardwareError{Reason: "QEMU version output is empty"}
	}
	if !validQEMUVersion(version) {
		return "", &HardwareError{Reason: "QEMU version output is invalid"}
	}
	return version, nil
}

type boundedOutput struct {
	bytes    []byte
	limit    int
	overflow bool
}

func (w *boundedOutput) Write(value []byte) (int, error) {
	remaining := w.limit - len(w.bytes)
	if remaining > 0 {
		w.bytes = append(w.bytes, value[:min(remaining, len(value))]...)
	}
	if len(value) > remaining {
		w.overflow = true
	}
	return len(value), nil
}

func compactASCIIJSON(value any) (string, error) {
	var output bytes.Buffer
	encoder := json.NewEncoder(&output)
	encoder.SetEscapeHTML(false)
	if err := encoder.Encode(value); err != nil {
		return "", err
	}
	encoded := hardwareASCIIJSON(output.Bytes())
	return strings.TrimSuffix(string(encoded), "\n"), nil
}

func hardwareASCIIJSON(encoded []byte) []byte {
	output := make([]byte, 0, len(encoded))
	for len(encoded) > 0 {
		r, size := utf8.DecodeRune(encoded)
		if r < utf8.RuneSelf {
			output = append(output, byte(r))
		} else if r <= 0xffff {
			output = fmt.Appendf(output, `\u%04x`, r)
		} else {
			first, second := utf16.EncodeRune(r)
			output = fmt.Appendf(output, `\u%04x\u%04x`, first, second)
		}
		encoded = encoded[size:]
	}
	return output
}

func validQEMUVersion(version string) bool {
	if version == "" || !utf8.ValidString(version) || len([]byte(version)) > qemuVersionLimit {
		return false
	}
	for _, value := range version {
		if !unicode.IsPrint(value) {
			return false
		}
	}
	return true
}

func firstPythonTextLine(value string) (string, bool) {
	if value == "" {
		return "", false
	}
	for index, character := range value {
		switch character {
		case '\n', '\v', '\f', '\r', 0x1c, 0x1d, 0x1e, 0x85, 0x2028, 0x2029:
			return value[:index], true
		}
	}
	return value, true
}

func equalStrings(left, right []string) bool {
	if len(left) != len(right) {
		return false
	}
	for index := range left {
		if left[index] != right[index] {
			return false
		}
	}
	return true
}

func requireHardwareJSONEOF(decoder *json.Decoder) error {
	var trailing any
	if err := decoder.Decode(&trailing); !errors.Is(err, io.EOF) {
		return errors.New("trailing JSON data")
	}
	return nil
}

func malformedHardwareManifest() error {
	return &HardwareError{Reason: "generation hardware manifest is malformed"}
}

func malformedHardwareManifestWith(err error) error {
	return &HardwareError{Reason: "generation hardware manifest is malformed", Err: err}
}
